File: array/129.cpp
#include <bits/stdc++.h>

using namespace std;

using pi = pair<int, int>;

int minSetSize(vector<int>& arr) {
    unordered_map<int, int> um;
    set<pair<int, int>, function<bool(const pi &,const pi&)>> l([](const pi &a,const pi &b){
        if(a.second == b.second)
            return a.first < b.first;
        return a.second > b.second;
    });

    int cont = 0;
    int cantKeysProccessed = 0;
    int n = arr.size();

    for(auto num: arr)
        um[num]++;

    l.insert(um.begin(), um.end());

    for(auto ptr: l) {
        cantKeysProccessed++;
        cont += ptr.second;

        if(n - cont <= int(n / 2))
            break;
    }

    return cantKeysProccessed;
}


int main() {
    vector<int> v {1,2,2,2,3,3,4,4,5};
    cout<<minSetSize(v);

    return 0;
}

File: array/129.py
from typing import List
from collections import defaultdict


def getSetSize(l: List[tuple[int, int]], n: int):
    cont = 0
    cant_nums_proccessed = 0

    for _, cant in l:
        cont += cant
        cant_nums_proccessed += 1

        if n - cont <= n / 2:
            break

    return cant_nums_proccessed


def minSetSize(arr: List[int]) -> int:
    freq = defaultdict(int)
    n = len(arr)

    for num in arr:
        freq[num] += 1

    l = sorted(freq.items(), key=lambda x: x[1], reverse=True)
    r = sorted(freq.items(), key=lambda x: x[1])

    sum_left = getSetSize(l, n)
    sum_right = getSetSize(r, n)

    return min(sum_left, sum_right)


arr = [3, 3, 3, 3, 5, 5, 5, 2, 2, 7]
print(minSetSize(arr))
